src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -251,128 +244,145 @@ private static string ComputeHash(string data, string algorithm)
+                while (currentIndex < data.Length && data[currentIndex] == ' ')
                 {
-                    temp++;
+                    currentIndex++;
                 }
 
                 // Start parsing key
-                while (*temp != '=')
+                int start = currentIndex;
+
+                // Parse till '=' is encountered marking end of key.
+                // Key cannot contain whitespace, break if whitespace is found.
+                while (currentIndex < data.Length && data[currentIndex] != '=' && data[currentIndex] != ' ')
                 {
-                    // Key cannot have whitespace
-                    if (*temp == ' ')
-                        break;
+                    currentIndex++;
+                }
 
-                    sb.Append(*temp);
-                    temp++;
+                if (currentIndex == data.Length)
+                {
+                    // Key didn't terminate with '='
+                    parsedIndex = currentIndex;
+                    return null;
                 }
 
-                // Skip trailing whitespace and '='
-                while (*temp == ' ' || *temp == '=')
+                // Record end of key.
+                int length = currentIndex - start;
+                if (data[currentIndex] == ' ')
                 {
-                    temp++;
+                    // Key parsing terminated due to whitespace.
+                    // Parse till '=' is found.
+                    while (currentIndex < data.Length && data[currentIndex] == ' ')
+                    {
+                        currentIndex++;
+                    }
+
+                    if (currentIndex == data.Length || data[currentIndex] != '=')
+                    {
+                        // Key is invalid.
+                        parsedIndex = currentIndex;
+                        return null;
+                    }
                 }
 
-                // Set the ref p to temp
-                p = temp;
+                // Skip trailing whitespace and '='
+                while (currentIndex < data.Length && (data[currentIndex] == ' ' || data[currentIndex] == '='))
+                {
+                    currentIndex++;
+                }
 
-                return StringBuilderCache.GetStringAndRelease(sb);
+                // Set the parsedIndex to current valid char.
+                parsedIndex = currentIndex;
+                return data.Substring(start, length);
             }
 
-            private unsafe string GetNextValue(ref char* p)
+            private string GetNextValue(string data, int currentIndex, out int parsedIndex)
             {
-                // It is generally cheaper to change a local and then write back to ref at the end
-                // rather than updating the ref on each operation.
-                char* temp = p;
-
-                StringBuilder sb = StringBuilderCache.Acquire();
-
                 // Skip leading whitespace
```

Comment:
Trailing whitespace (including whitespace after the '=') looks to have already been skipped in getting the key, right?  So is there anything further to skip here?

Reply:
Yeah technically it should be skipped, just added the check here, since this method is independent of getkey, and didn't want to take dependency of validation check in another method. Can convert this to debug.assert